x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/ShrinkActionTests.java

```diff
@@ -66,18 +66,19 @@ protected ShrinkAction createTestInstance() {
 
     static ShrinkAction randomInstance() {
         if (randomBoolean()) {
-            return new ShrinkAction(randomIntBetween(1, 100), null);
+            return new ShrinkAction(randomIntBetween(1, 100), null, randomBoolean());
         } else {
-            return new ShrinkAction(null, ByteSizeValue.ofBytes(randomIntBetween(1, 100)));
+            return new ShrinkAction(null, ByteSizeValue.ofBytes(randomIntBetween(1, 100)), randomBoolean());
         }
     }
 
     @Override
     protected ShrinkAction mutateInstance(ShrinkAction action) {
+        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink() == false;
         if (action.getNumberOfShards() != null) {
-            return new ShrinkAction(action.getNumberOfShards() + randomIntBetween(1, 2), null);
+            return new ShrinkAction(action.getNumberOfShards() + randomIntBetween(1, 2), null, allowWriteAfterShrink);
         } else {
-            return new ShrinkAction(null, ByteSizeValue.ofBytes(action.getMaxPrimaryShardSize().getBytes() + 1));
+            return new ShrinkAction(null, ByteSizeValue.ofBytes(action.getMaxPrimaryShardSize().getBytes() + 1), allowWriteAfterShrink);
         }
```

Comment:
I think we usually try to only change one field, to make sure every single field is covered in `equals`, `hashCode`, etc. I see that the original code already changed both, but while we're here, we might as well fix it.

Reply:
Ahh, yeah that makes sense. Thank for the review!

Reply:
Looks like it was actually not changing both fields at once in the original code. The format of this method is the bit different from the more common pattern of using a switch with case for each field, as only one of the first two parameters can be non-null. I've fixed it now to update only one field at a time, though the if statement is a little wonky looking with the `randonInBetween(0, 2) == 0`, as I didn't want to use randBoolean() and spent 50% of the time toggling a single boolean field.

Reply:
Ah sorry I must have read the test a bit too quickly. What about doing:
```java
        Integer numberOfShards = action.getNumberOfShards();
        ByteSizeValue maxPrimaryShardSize = action.getMaxPrimaryShardSize();
        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink();

        switch (randomInt(2)) {
            case 0 -> numberOfShards = randomValueOtherThan(numberOfShards, () -> randomIntBetween(1, 100));
            case 1 -> maxPrimaryShardSize = randomValueOtherThan(
                maxPrimaryShardSize,
                () -> ByteSizeValue.ofBytes(randomIntBetween(1, 100))
            );
            case 2 -> allowWriteAfterShrink = allowWriteAfterShrink == false;
        }
        return new ShrinkAction(numberOfShards, maxPrimaryShardSize, allowWriteAfterShrink);
```
That way it still follows the "common pattern" and is still correct. The only downside is that we don't have a mutation that changes which of the first two fields is null/non-null, but I don't think that's too big of an issue as the `randomInstance()` method should take care of randomizing that. What do you think?

Reply:
Unfortunately, that could produce an invalid object. For example if numberOfShards is  originally non-null and maxPrimaryShardSize is null, then if `case 1` is selected, both values will be non-null, which is invalid. (the only reason I know this is because I wrote something similar and then broke the test!)

Reply:
Ahh of course 🤦🏻.
Then this is my last suggestion 😉. I've seen this in several places too. It's a bit more verbose than what you have now, but, since it aligns more with the common pattern, I think it'll be easier to read.
```java
        Integer numberOfShards = action.getNumberOfShards();
        ByteSizeValue maxPrimaryShardSize = action.getMaxPrimaryShardSize();
        boolean allowWriteAfterShrink = action.getAllowWriteAfterShrink();

        switch (randomInt(2)) {
            case 0 -> {
                numberOfShards = randomValueOtherThan(numberOfShards, () -> randomIntBetween(1, 100));
                maxPrimaryShardSize = null;
            }
            case 1 -> {
                maxPrimaryShardSize = randomValueOtherThan(
                    maxPrimaryShardSize,
                    () -> ByteSizeValue.ofBytes(randomIntBetween(1, 100))
                );
                numberOfShards = null;
            }
            case 2 -> allowWriteAfterShrink = allowWriteAfterShrink == false;
        }
        return new ShrinkAction(numberOfShards, maxPrimaryShardSize, allowWriteAfterShrink);
```

Since this is less of a clear "improvement", I'll let it up to you which way you want to go :)

Reply:
I like it :+1:  Seems worthwhile to match the common pattern.